Unix.environment on Windows: use _wenviron

nojb opened the pull request:
One more rebase fix following #1200.

The existing `Unix.environment` in `trunk` uses the native Win32 API `GetEnvironmentStrings` which works OK except that its result can sometimes get out of sync with the result of the "POSIX" API `_wgetenv` used for `Unix.{get,put}env` (see [MPR#4499](https://caml.inria.fr/mantis/view.php?id=4499) for details).

The mixup ocurred because at some point we had switched to using the native Win32 API in all three functions but later decided to leave that change for later.

This patch reverts `Unix.environment` to read from the global variable `_wenviron`, similar to the analogous implementation for Unix.

nojb committed:
Unix.environment on Windows: use _wenviron

alainfrisch commented on otherlibs/win32unix/envir.c:
```diff
@@ -21,23 +21,26 @@
 #include <caml/osdeps.h>
 
 #include <Windows.h>
+#include <Stdlib.h>
```
stdlib (lowercase) is more standard (also for windows.h, btw) in our codebase.

dra27 replied:
FWIW, an quick check in the museum shows that Microsoft use Windows.h and stdlib.h (or, of course, STDLIB.H if you're very, very old...)

However, the MinGW headers use windows.h so it's anyone's guess really... let's hope no one is crazy enough to try running the kernel in case sensitive mode :smile:

alainfrisch commented on otherlibs/win32unix/envir.c:
```diff
@@ -21,23 +21,26 @@
-    v = caml_copy_string_of_utf16(p);
-    Store_field(result, i ++, v);
+  if (_wenviron != NULL) {
+    nbr = 0;
```
Cannot one use caml_alloc_array as caml_copy_string_array does (passing caml_copy_string_of_utf16 instead)?

alainfrisch commented:
Could you create a test that would fail before this PR is merged?

nojb commented:
The point of the PR is that it **should** be possible - except that today I can't seem to reproduce the failure locally, mmmph... Will investigate more, see if I can figure out what is going on.

In any case, it seems weird to use the native Win32 API for just one of the calls instead of all three.

alainfrisch commented:
At least for the ANSI version, my understanding of MPR4499 is that putenv indeed sends the new value to the "real "environment" (the one manipulated through GetEnvironmentVariable and co). If this is also the case for the W version of putenv, everything would be ok (although I understand it is not very nice to mix the two kinds of functions).  It would actually be a bit "better" since this would allow to see changes performed through SetEnvironmentVariable (e.g. from .NET).  But yes, it would actually be confusing to see changes that are invisible to getenv...

dra27 commented:
Ugh, it's one of many ugly sections of the Microsoft C runtime! It feels somewhat icky because _environ and _wenviron are deprecated (however, MSVC also provides no enumeration functions, as far as I'm aware, so go figure!).

OCaml is supposed to sit atop a C runtime, even if it's an often irritating one, so I agree that correcting this one function is the way to go.

nojb commented:
Do I need to add an entry to Changes?

alainfrisch commented:
> Do I need to add an entry to Changes?

Perhaps mention it together with the entry for #1200.

nojb committed:
Add Changes entry

nojb commented:
Ah yes, of course!

olleolleolle commented on otherlibs/win32unix/envir.c:
```diff
@@ -21,23 +21,26 @@
-  wchar_t * envp, * p;
-  int size, i;
+  CAMLparam0 ();
+  CAMLlocal2 (v, result);
```
Hi!

Are the spaces after the function names introduced by design?

It's on line 28, 29, 36.

nojb replied:
Thanks for the comment; and not, not really. Will fix it and give AppVeyor an excuse to wake up again.

xavierleroy commented:
@alainfrisch suggested a possibly simpler implementation.  Please tell why it would not work.

nojb commented:
Yes, sorry about that, I forgot to post an answer here.  The reason was that `caml_copy_array` is not "polymorphic" enough, it has signature
```
value caml_alloc_array(value (*funct)(char const *), char const ** arr)
```
(we need a version with `wchar_t` in lieu of `char`). It is tempting to change `char *` to `void *` in this signature, but then the `funct` argument will have type `value (*funct)(void const *)` and will not accept something less general like `value caml_copy_string(char const *)`.

Anyone see a way around this?

nojb commented:
Of course I can just cast the argument to use it (in fact I already used this once: https://github.com/ocaml/ocaml/blob/trunk/byterun/sys.c#L362).  Shall I do that?

nojb commented:
OK, absent any explicit decision I went ahead and added a commit using `caml_alloc_array` (with a cast), which looks simpler.

gasche commented:
Is it possible/practical to add a regression test for the bug you are fixing?

nojb committed:
Update testsuite

- Add lib-unix/win-env/test_env.ml, a test that fails in trunk and passes in
  this PR.

- Add lib-unix/win-env/test_env2.ml: a test for MPR#4499 which is closely
  related to this PR.  Currently disabled as it fails.

- Added a new test in win-unicode/mltest.ml to test non-ASCII characters
  returned by Unix.environment (Unix.{getenv,putenv} were already being tested).

nojb commented:
Yes, it is!  I added a test (`test_env.ml`) which fails on `trunk` and passes on this branch.  I also added a second (disabled) test `test_env2.ml` which fails due to the closely related [MPR#4499](https://caml.inria.fr/mantis/view.php?id=4499). I am preparing a PR for after 4.06 to fix the second one.

gasche commented:
Is someone willing to review and approve the PR? ( @alainfrisch, @dra27 ? )

alainfrisch commented on testsuite/tests/win-unicode/test_env.ml:
```diff
@@ -0,0 +1,30 @@
+let foo = "FOO"
+
+let () =
+  set_environment_variable foo "BAR";
```
What about storing a non-ASCII text here (utf-8 encoded), to also check that Unicode characters are properly passed back?

alainfrisch commented on testsuite/tests/win-unicode/test_env2.ml:
```diff
@@ -0,0 +1,13 @@
+let foo = "FOO"
+
+let () =
+  set_environment_variable foo "BAR";
```
It's good to have this test (disabled), but perhaps indicates somehow that it is related to MPR#4499, not the current PR.

Also, it might be good to use non-ASCII characters in the string.

alainfrisch commented:
LGTM (modulo a minor comment about the disabled test).  Waiting for the CI green lights.

nojb commented:
Rebased and amended as suggested:
- Added a test to check non-ASCII characters being returned by `Unix.environment` to `win-unicode/mltest.ml` (`Unix.{getenv,putenv}` were already being tested)
- Added a new test `lib-unix/win-env/test_env.ml` for this PR (fails on `trunk`)
- Added a new test `lib-unix/win-env/test_env2.ml`, currently disabled (with a comment explaining why), that tests MPR#4499 which is closely related to the present PR.

gasche commented:
( @alainfrisch does your comment count as approval? Let's follow the process as it was suggested. )

alainfrisch commented:
@damiendoligez @gasche I'd like to cherry-pick this to 4.06.  Do you agree?

gasche commented:
Yes, I think it's good to have it in the first release with #1200. I would encourage you to include it in 4.06.

alainfrisch commented:
Cherry-picked to 4.06 as 990845b64.
